### PR TITLE
Add build step for producing Android Archive (AAR)

### LINKED ISSUE #136: Add build step to create Android archive (AAR) file
@ademar111190 as you said in the past it would be easier to Android dev to work with rather than configure the binaries.

### PR DESCRIPTION
**Pull request by cnixbtc**
Resolves #136.

This approach loosely [follows bdk's Android build step](https://github.com/bitcoindevkit/bdk-ffi/blob/master/.github/workflows/publish-android.yaml).

This adds a build step to crate an Android Archive (AAR). This works by setting up an Android Studio library project in `bindigns-android` and then copy the compiled libraries and Kotlin bindings into that project during the build step. Then we can build the project using Gradle which will produce an AAR as build output.

Most of the files in `bindings-android` are just default Android Studio config files.

### TL;DR

```
make bindings-android
```

will generate an AAR in `ffi/android/lib-release.aar`.

### Next Steps

As we do with the Swift package, I think it would make sense to publish the library somewhere. I don't have too much Android experience but it seems like [publishing a Maven repository via Sonatype OSSRH is the way to go](https://getstream.io/blog/publishing-libraries-to-mavencentral-2021/) (which is also what bdk does).

Doing that requires setting up the repository and configuring gradle to publish to the repository. We can then setup a GitHub action that runs the gradle publish task everytime we tag a release in _breez-sdk_. I'll have a closer look at how exactly bdk handles this and will come up with an approach in a separate PR.

### CODE REVIEW

**Comment by ok300** [comment]
If we host the Maven repo on a 3rd party website, we should make sure to include signed hashes[^1] for our artifacts.

If we want to host the repo ourselves, there are tools like Reposilite[^2]. Its a simple Maven repo implementation available as docker image.

[^1]: https://docs.gradle.org/current/userguide/dependency_verification.html#sec:signature-verification
[^2]: https://reposilite.com

**Review by roeierez — COMMENTED** [review]
Gradle approach sounds reasonable. Dropped a comment regarding the package name. Other than that looks good to me.

File: libs/sdk-bindings/makefile [hunk 1]
```diff
@@ -66,22 +65,30 @@ bindings-swift: ios-universal darwin-universal
...
 i686-linux-android: $(SOURCES) ndk-home
 	cargo ndk -t i686-linux-android -o ffi/kotlin/jniLibs build --release	
 
 x86_64-linux-android: $(SOURCES) ndk-home
-	cargo ndk -t x86_64-linux-android -o ffi/kotlin/jniLibs build --release	
+	cargo ndk -t x86_64-linux-android -o ffi/kotlin/jniLibs build --release
+
+bindings-android: android
+	cp -r ffi/kotlin/jniLibs bindings-android/lib/src/main
+	sed 's/package_name = "breez_sdk"/package_name = "technology.breez"/' uniffi.toml > uniffi.android.toml
```
[the remark this answers is not in the document]
Comment (cnixbtc) [reply libs/sdk-bindings/makefile:86]: You're right. Originally, I was thinking that `technology.breez` was closer to the convention on Android / in the Java world but I didn't want to break the other integrations so I changed it just for Android. But thinking about it now, I believe consistency trumps convention. Will change this so `breez_sdk` is the package name on Android as well.

**Comment by cnixbtc** [comment]
> If we host the Maven repo on a 3rd party website, we should make sure to include signed hashes[1](https://github.com/breez/breez-sdk/pull/184#user-content-fn-1-a755983e48d715a7a7dfb9f98ebf50e6) for our artifacts.
>
> If we want to host the repo ourselves, there are tools like Reposilite[2](https://github.com/breez/breez-sdk/pull/184#user-content-fn-2-a755983e48d715a7a7dfb9f98ebf50e6). Its a simple Maven repo implementation available as docker image.

Fully agree! Is there any preferences from your end on self-hosting vs. going the hosted route?  cc @roeierez  

In my experience self-hosting is generally the "more beautiful" way to do things but in a commercial setting only really makes sense if you have resources to look after the stuff and ensure uptime etc. I'm speaking generally here not specific to Maven repositories (I have no experience with those).

Let me know what you guys think!

**Comment by cnixbtc** [comment]
I just tried updating the package name but when trying to use the resulting AAR in an example project, I'm getting an error seemingly related to JNA:

```
java.lang.UnsatisfiedLinkError: Error looking up function 'ffi_breez_sdk_6313_EventListener_init_callback': undefined symbol: ffi_breez_sdk_6313_EventListener_init_callback
```

I tried reverting the changes I made but couldn't get rid of the error. Could've sworn that when I opened the PR everythign worked, though. Will need to investigate further.

**Comment by roeierez** [comment]
> > If we host the Maven repo on a 3rd party website, we should make sure to include signed hashes[1](https://github.com/breez/breez-sdk/pull/184#user-content-fn-1-a755983e48d715a7a7dfb9f98ebf50e6) for our artifacts.
> > If we want to host the repo ourselves, there are tools like Reposilite[2](https://github.com/breez/breez-sdk/pull/184#user-content-fn-2-a755983e48d715a7a7dfb9f98ebf50e6). Its a simple Maven repo implementation available as docker image.
> 
> Fully agree! Is there any preferences from your end on self-hosting vs. going the hosted route? cc @roeierez
> 
> In my experience self-hosting is generally the "more beautiful" way to do things but in a commercial setting only really makes sense if you have resources to look after the stuff and ensure uptime etc. I'm speaking generally here not specific to Maven repositories (I have no experience with those).
> 
> Let me know what you guys think!

@yaslama is responsible for the IT, what do you think Akiba?

File: libs/sdk-bindings/makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (roeierez) [reply libs/sdk-bindings/makefile:86]: Also @cnixbtc we want to change the RN package to be dependent on this aar file (instead of requiring developers to build the binaries themselves). And for that it must be the same package used in the RN bundle.

[the remark this answers is not in the document]
Comment (cnixbtc) [reply libs/sdk-bindings/makefile:86]: Makes sense! I'm still looking into what seems to be a runtime linker issue (see comment further down). I'll update this PR with a fix for that and update the package name before we merge it.

**Comment by dlggr** [comment]
I've reverted the package name back to `breez_sdk`. The linker error seems to be gone as well. I still had to keep `technology.breez` [as the package name in the manifest file](https://github.com/breez/breez-sdk/pull/184/files#diff-e4fe966c8aa91cd76c3b7105f70e25620e1d27081035de834460d308e5c84a31R2) since gradle wouldn't let me build the project otherwise (it requires a valid Android package name in the Manifest file).

The resulting AAR will include two packages:

Screenshot 2023-05-02 at 23 57 40

- `technology.breez`
  - This only includes a `BuildConfig.class` autogenerated during the build containing some metadata such as a _DEBUG_ flag
- `breez_sdk`
  - All the Breez SDK Kotlin bindings

The Kotlin bindings in the `breez_sdk` package can then be used like this:


``` kotlin
import breez_sdk.*

var seed = mnemonicToSeed(/*...*/)
```

---

This PR is ready for testing/review again. I'd really appreciate someone giving the generated AAR file a go to make sure the linking error is really gone and it's not related to something local on my machine.

**Comment by dlggr** [comment]
As for publishing to a Maven repo: I'd say we do this in a separate PR.

@yaslama is already setting up a self-hosted repo. I will then add configuration to push to that repo in a separate PR.

**Comment by cnixbtc** [comment]
I've reverted the package name back to `breez_sdk`. The linker error seems to be gone as well. I still had to keep `technology.breez` [as the package name in the manifest file](https://github.com/breez/breez-sdk/pull/184/files#diff-e4fe966c8aa91cd76c3b7105f70e25620e1d27081035de834460d308e5c84a31R2) since gradle wouldn't let me build the project otherwise (it requires a valid Android package name in the Manifest file).

The resulting AAR will include two packages:

Screenshot 2023-05-02 at 23 57 40

- `technology.breez`
  - This only includes a `BuildConfig.class` autogenerated during the build containing some metadata such as a _DEBUG_ flag
- `breez_sdk`
  - All the Breez SDK Kotlin bindings

The Kotlin bindings in the `breez_sdk` package can then be used like this:


``` kotlin
import breez_sdk.*

var seed = mnemonicToSeed(/*...*/)
```

---

This PR is ready for testing/review again. I'd really appreciate someone giving the generated AAR file a go to make sure the linking error is really gone and it's not related to something local on my machine.

---

As for publishing to a Maven repo: I'd say we do this in a separate PR.

@yaslama is already setting up a self-hosted repo. I will then add configuration to push to that repo in a separate PR.

**Review by roeierez — APPROVED** [review]
